File: algorithms/Koko Eating Bananas/Koko Eating Bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        
        int mx=1000000001;
        
        
        int st=1;
        
        
        while(mx>st)
        {
            
            int k = ((mx+st)/2);
            
            int sum=0;
            
            for(int i =0 ;i < piles.size() ; i++)
            {
                sum+=ceil(1.0 *piles[i]/k);
            }
            
            if(sum>h)
            {
                st=k+1;
            }
            else
            {
                mx=k;
            }
            
        }
        
        
        return st;
        
        
        
    }
};

File: algorithms/K/Koko Eating Bananas/Koko Eating Bananas.js
var minEatingSpeed = function(piles, h) {
    let low = 1, high = Math.max(...piles)
    let minBanana = Infinity
    while(low <= high){
        const banana = Math.floor((high - low) / 2) + low
        const hours = piles.reduce((a, c) => a + Math.ceil(c / banana), 0)
        if(hours <= h){
            minBanana = banana
            high = banana - 1
        }
        else low = banana + 1
    }
    return minBanana
};
// time O(nlog(max(banana)), space O(1) 
